FEA Add support for missing values in tree estimators with `criterion="absolute_error"` by greatly simplifying the logic

cakedev0 opened the pull request:
This PR refactors how missing values are handled in trees by:
- removing missing-values handling from `Criterion` subclasses 
- making it the responsability of the splitter & partitionner only

This greatly simplifies the logic and unlocks for free the support of missing values for MAE trees.

#### Reference Issues/PRs

Accidentally fixes #32870 

Otherwise, I looked but I didn't find any issue requesting the support of missing values with `criterion="absolute_error"`. Just this recent comment from @ogrisel in this PR: https://github.com/scikit-learn/scikit-learn/pull/32100#pullrequestreview-3387374355. Maybe it's because MAE trees were just too slow in sklearn before PR #32100, so they aren't used a lot for now.

#### What does this implement/fix? Explain your changes.

Currently, a part of the missing values support is done by each subclass of `Criterion`. I believe it's not a great design because:
- `Criterion` is "X-blind", it's not aware of X values. It just looks at `y` and `sample_weights` in the order defined by the sorted indices (`sample_indices`). It never looks at X values. But somehow, by making it handle missing values, it does have some dealing with X values. Why not just use the ordering of `sample_indices` to take into account missing values? Like what we do for any other value (even inf/-inf for instance).
- It requires each criterion to implement nan-handling logic.

So, to the question "Why not just use the ordering of `sample_indices` to take into account missing values? " my answer is :"yes, let's just do that". The result is removing 200 lines from `_criterion.pxy` while not increasing the complexity of the splitter and the partitioner (actually, it also simplifies a bit the splitter).

#### Any other comments?

I think it might unlock adding the support for missing values + monotonic constraints without too much efforts, but I haven't look into it yet.

It might also simplify a bit the support for missing values + sparse, but this would still be hard (probably not worth the effort).

cakedev0 committed:
first draft: compilation ok

cakedev0 committed:
fixed seg fault bug

cakedev0 committed:
some tests with decision tree and missing values are passing

cakedev0 committed:
WIP

cakedev0 committed:
WIP

cakedev0 committed:
fixed the silly mistake that was causing seg faults & infinite loops

cakedev0 committed:
AE is now supported

cakedev0 committed:
Fix last major bug - in random splits

cakedev0 committed:
cleanup

cakedev0 committed:
cleanup prints

cakedev0 committed:
cleanup unsued var

cakedev0 committed:
cleanup

cakedev0 committed:
attempt at fixing CI failing for 32-bits systems

cakedev0 committed:
removed checks on criterion; included abs err in more tests

cakedev0 committed:
removed a line added for exp I forgot to remove

cakedev0 committed:
WIP debuggin 32-bits tests

cakedev0 commented:
Note: at this day, tests passes on my laptop and most CI unit tests pipelines are successful. But some are failing, I managed to reproduce one of the failing pipelines locally using a Docker image. I still need to find the bug though.

cakedev0 committed:
fixed silly bug that took way too long to find

cakedev0 committed:
WIP debugging

cakedev0 committed:
tests are passing? Or am I too tired?

cakedev0 committed:
Probably fixed all the bugs

cakedev0 committed:
Removed all debug prints

cakedev0 committed:
cleanup; comments; unit test for swap slices

cakedev0 committed:
changelog

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
simpler next_p logic

adam2392 commented:
Let's keep this in draft mode until we merge #32100. Ping here for a review after the initial PRs are sorted out

cakedev0 committed:
test for optimal splits

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
comments

cakedev0 committed:
added friedman mse

cakedev0 committed:
update next_p docstring

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
minor cleanup

cakedev0 committed:
fix

cakedev0 committed:
test sparse; clean-up

cakedev0 committed:
added categorical

cakedev0 committed:
remove criterion limitation in doc

cakedev0 committed:
changelog for the by-product fix

cakedev0 committed:
Merge branch 'main' into tree-simpler-missing

cakedev0 committed:
Merge branch 'main' into testing-split

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
fix linting

cakedev0 committed:
import order

ogrisel commented on sklearn/tree/_classes.py:
```diff
@@ -1991,10 +1977,6 @@ def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
         # XXX: nan is only supported for dense arrays, but we set this for the
```
```suggestion
        # XXX: nan values are only accepted in dense arrays, but we set this for
```

ogrisel reviewed (commented):
I don't have the time to fully review this today, but here is some cosmetic feedback from a first glance at the diff.

cakedev0 committed:
Apply suggestions from code review

comments typos

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

cakedev0 committed:
Update sklearn/tree/_classes.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

cakedev0 committed:
Merge branch 'main' into tree-simpler-missing

cakedev0 committed:
remove unsed attribute

cakedev0 committed:
Merge branch 'main' into tree-simpler-missing

cakedev0 committed:
Merge branch 'main' into testing-split

cakedev0 committed:
Merge branch 'testing-split' of github.com:cakedev0/scikit-learn into testing-split

cakedev0 committed:
Merge branch 'main' into testing-split

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

cakedev0 committed:
add non-reg test for poisson + missing values

lorentzenchr reviewed (commented):
A start of a review...

cakedev0 commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -2581,12 +2586,12 @@ def test_missing_values_best_splitter_missing_both_classes_has_nan(criterion):
 
 
-@pytest.mark.parametrize("sparse_container", [None] + CSR_CONTAINERS)
+@pytest.mark.parametrize("sparse_container", CSR_CONTAINERS)
```
This test is `test_missing_value_errors`, and checks `ValueError` is raised when X contains missing values but it's not supported. Now it's supported for all cases when dense X.

... well actually, it's not supported for dense X when monotonic constraints are given, but this case is already tested in the file `test_monotonic_tree.py`.

lorentzenchr replied:
ok

cakedev0 committed:
test all criteria

cakedev0 committed:
Apply suggestion from @lorentzenchr

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

cakedev0 committed:
slight change change reformulation

cakedev0 committed:
avoid void *

cakedev0 commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -1853,16 +1853,3 @@ def test_missing_value_is_predictive(Forest):
-
-
-@pytest.mark.parametrize("Forest", FOREST_REGRESSORS.values())
-def test_non_supported_criterion_raises_error_with_missing_values(Forest):
```
Why not, but I would say that there are already enough tests.

If another reviewer +1 your comment, I'll do it.

adam2392 replied:
Ah okay. Misunderstood your og comment. 

In that case, I do agree with @lorentzenchr  

We ideally want a unit test (to replace this one) to test the MAE with missing values with RF. Having a test for MAE with missing values with a single decision tree is good. But we also do indeed still need one for RF.

cakedev0 committed:
fix & improve test_py_swap_array_slices_random

lesteve commented:
cc @adam2392 in case you have some time and have some comment on the high-level approach?

> This PR refactors how missing values are handled in trees by:
> - removing missing-values handling from Criterion subclasses
> - making it the responsability of the splitter & partitionner only

adam2392 commented:
I should have more time next week. Last push before the holidays at work. Feel free to continue pinging me.

cakedev0 committed:
comment in test: more details about y > 0 and poisson criterion

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 reviewed (commented):
LGTM, minor comments about tests. I haven't really reviewed the part related to fixing `class_weight = "balanced"`.

lorentzenchr commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -2785,21 +2785,27 @@ def test_regression_tree_missing_values_toy(Tree, X, criterion, global_random_se
+    y = np.arange(1, 7)
+    # y needs to be > 0 for this test to pass with poisson criterion
+    # Poisson criterion allows multiple optimal splits when y[i] = 0 because
+    # the impurity calculation involves log(y[i]), which is undefined at 0.
```
This is not entirely correct. And for me, it does not relate to the next sentence.

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

lorentzenchr commented on doc/whats_new/upcoming_changes/sklearn.tree/32119.fix.rst:
```diff
@@ -0,0 +1,4 @@
+- Fix calculation of nodes impurity in :class:`tree.DecisionTreeRegressor`
+  and :class:`ensemble.RandomForestRegressor` when missing values are
+  present.
```
Could you be a bit more specific when this error occured? For all criteria? Allways when missing values are present?

lorentzenchr commented on sklearn/tree/tests/test_utils.py:
```diff
@@ -0,0 +1,29 @@
+        """
+        arr[start:end] = np.concatenate([arr[start + n : end], arr[start : start + n]])
+
+    rng = np.random.default_rng(0)
```
Better use the global random state fixture instead of the seed 0.

lorentzenchr reviewed (commented):
Another bit.

cakedev0 committed:
Forests: testing missing x criteria

cakedev0 committed:
trees: skip testing poisson in test_regression_tree_missing_values_toy

cakedev0 committed:
trees: explain poisson regression test

cakedev0 committed:
update changelog

cakedev0 committed:
improved test_py_swap_array_slices_random

cakedev0 committed:
Update doc/whats_new/upcoming_changes/sklearn.tree/32119.feature.rst

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

cakedev0 commented on sklearn/tree/tests/test_utils.py:
```diff
@@ -0,0 +1,29 @@
+        and array[start + n:end] while preserving the order
+        in the slices. Works for any itemsize.
+        """
+        arr[start:end] = np.concatenate([arr[start + n : end], arr[start : start + n]])
```
Changed that. It still does a copy inside, but it's not the responsibility of the tester anymore :+1:

cakedev0 committed:
Merge branch 'main' into tree-simpler-missing

cakedev0 committed:
reworded comment in test

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

cakedev0 commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -3047,3 +3053,14 @@ def test_missing_values_and_constant_toy():
+
+    tree = DecisionTreeRegressor(criterion="poisson")
+    tree.fit(X, y)
+    # impurity should be non-negative:
```
Did that and merged with the explanation of the issue.

cakedev0 committed:
fix tests/CI

cakedev0 committed:
[all random seeds] test_py_swap_array_slices_random

adam2392 commented on sklearn/tree/_partitioner.pyx:
```diff
@@ -110,102 +126,85 @@ cdef class DensePartitioner:
-            p[0] + 1 < end_non_missing and
-            self.feature_values[p[0] + 1] <= self.feature_values[p[0]] + FEATURE_THRESHOLD
-        ):
+        - if self.missing_on_the_left: go over the p in [start + n_missing + 1, end[
```
```suggestion
        - if self.missing_on_the_left: go over the p in [start + n_missing + 1, end]
```
?

adam2392 replied:
Ah I see. I thought it was a typo :p. Yeah perhaps use ")" to indicate non-inclusive closed bracket.

cakedev0 replied:
```suggestion
        - if self.missing_on_the_left: go over the p in [start + n_missing + 1, end)
```

adam2392 commented on sklearn/tree/_partitioner.pyx:
```diff
@@ -110,102 +126,85 @@ cdef class DensePartitioner:
-            float32_t[::1] feature_values = self.feature_values
+            intp_t partition_end = self.end
+            intp_t* samples = &self.samples[0]
+            float32_t* feature_values = &self.feature_values[0]
```
Why do we need the pointers? Can we just pass to swap the references of self.features_values[0] and self.samples[0]

Or does that not work? It's been awhile since I've thought about the implications in cython :p

adam2392 commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -2769,7 +2769,8 @@ def test_deterministic_pickle():
 )
-@pytest.mark.parametrize("criterion", ["squared_error", "friedman_mse"])
+@pytest.mark.parametrize("criterion", sorted(set(REG_CRITERIONS) - {"poisson"}))
+# TODO: adapt this test for poisson criterion
```
Why not have this test work for poisson criterion here and now?

cakedev0 replied:
Well, this would allow us to merge another test into this one, so let's do it.

But I don't really know how to justify the change from `y = np.arange(6)` to `y = np.arange(1, 7)`... And honestly, I don't want to spend time finding the exact justification except that "poisson isn't really supposed to work with 0s" 😅

adam2392 commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -2769,7 +2769,8 @@ def test_deterministic_pickle():
+# TODO: adapt this test for poisson criterion
 def test_regression_tree_missing_values_toy(Tree, X, criterion, global_random_seed):
     """Check that we properly handle missing values in regression trees using a toy
     dataset.
```
The docstring of this unit-test should prolly get updated. It reads rn:

```
The regression targeted by this test was that we were not reinitializing the
    criterion when it comes to the number of missing values. Therefore, the value
    of the critetion (i.e. MSE) was completely wrong.
    This test check that the MSE is null when there is a single sample in the leaf.
```

adam2392 commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -3047,3 +3048,16 @@ def test_missing_values_and_constant_toy():
+
+    tree = DecisionTreeRegressor(criterion="poisson")
+    tree.fit(X, y)
+    assert (tree.tree_.impurity >= -1e10).all()
```
Can we just fold this test into the above `test_regression_tree_missing_values_toy`? In the parametrization, you can add a tolerance, and use the docstring/inline comments to help elaborate what's going on.

adam2392 reviewed (commented):
Left some comments. I am 95% sure the logic in Cython looks good to me. Given the unit-tests pass, that is good :)

cakedev0 committed:
minor edits of test docstring

cakedev0 committed:
merged new regression test into already existing test

cakedev0 commented on sklearn/tree/_partitioner.pxd:
```diff
@@ -60,7 +60,7 @@ cdef const float32_t FEATURE_THRESHOLD = 1e-7
 #         float64_t best_threshold,
 #         intp_t best_feature,
-#         intp_t n_missing,
+#         bint best_missing_go_to_left
```
Hum... maybe we could just pass `best_split` to this function, wdyt?

We're always calling it this way:
```python
            partitioner.partition_samples_final(
                best_split.pos,
                best_split.threshold,
                best_split.feature,
                best_split.missing_go_to_left
            )
```

cakedev0 committed:
rename in test

cakedev0 committed:
removed value initialization in partitioner

cakedev0 committed:
Update sklearn/tree/_partitioner.pyx

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

cakedev0 commented:
I benchmarked here a few months ago: https://github.com/scikit-learn/scikit-learn/pull/32119#issuecomment-3325355782

I've re-ran the benchmarks today, and also tried some different distribution for X, and it's the same between main and this branch.

<details>
<summary>Updated benchmark code</summary>

```Python
from time import perf_counter
import numpy
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier

d = 10
n_fit = 10
n_skip = 5

for Tree in [DecisionTreeRegressor, DecisionTreeClassifier]:
    for max_depth in [4, None]:
        n = 3_000_000 // d
        if max_depth is None:
            n //= 3
        for distrib in ['unif', 'geom']:
            dts = []
            for _ in range(n_fit):
                X = numpy.random.rand(n, d) if distrib == 'uniform' else numpy.random.geometric(0.2, size=(n * 3, d))
                X = X / X.max()
                y = numpy.random.rand(X.shape[0]) + X.sum(axis=1)
                if Tree is DecisionTreeClassifier:
                    y = y < numpy.median(y)
                for i in range(d):
                    X[i::(3 + i), i] = numpy.nan
                t = perf_counter()
                tree = Tree(max_features=d, max_depth=max_depth).fit(X, y)
                dts.append(perf_counter() - t)
            avg = numpy.mean(dts[n_skip:])
            std = numpy.std(dts[n_skip:])
            print(f"[X ~ {distrib}({n}, {d})] {Tree.__name__}(max_depth={max_depth}) : {avg:.2f} ± {std:.3f}s")
```

</details>

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Add translation friedman_mse -> squared_error and warning

cakedev0 committed:
Merge branch 'fix/broken_deprecation_fmse' into tree-simpler-missing

cakedev0 committed:
filter warnings

ogrisel commented on sklearn/tree/tests/test_utils.py:
```diff
@@ -0,0 +1,29 @@
+        expected = swap_slices_np(arr, start, end, split)
+
+        _py_swap_array_slices(arr, start, end, split)
+        np.testing.assert_array_equal(arr, expected)
```
and invalid inputs as well, possibly in another test.

cakedev0 replied:
I'd prefer avoid handling this kind of cases and avoid testing them because it's a private function, and the private functions don't do this kind of input validation in the trees.

For instance `sort`, you can give a `n` bigger than the arrays and it will segfault, but it's not something we guard against. So I don't see why `swap_array_slices` should do that.

ogrisel commented on sklearn/tree/_partitioner.pyx:
```diff
@@ -214,57 +212,32 @@ cdef class DensePartitioner:
         intp_t best_pos,
```
Why do we need to pass `best_pos` here?

ogrisel reviewed (commented):
Here is the feedback from my first review pass. I agree that the new design is an overall improvement.

I am a bit worried that our existing tests are a bit weak but maybe we can leave a refactoring / extension of the test suite for a dedicated PR (follow-up or concurrent to this one).

cakedev0 committed:
Apply suggestions from code review

Improve docstrings

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

cakedev0 committed:
Remove unnecessary check

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

ogrisel committed:
TST: compare the single split impurity to pure Python brute force solution

ogrisel committed:
[all random seeds] test_impurity_values_on_toy_data

cakedev0 committed:
minor refactor to remove best_pos argument from partition_final

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

cakedev0 committed:
adressed minor PR comments

lorentzenchr commented on sklearn/ensemble/_forest.py:
```diff
@@ -1928,6 +1928,16 @@ def __init__(
             max_samples=max_samples,
         )
 
+        if isinstance(criterion, str) and criterion == "friedman_mse":
```
What does this have to do with missing values?

lorentzenchr reviewed (commented):
Small question from the sideline.

cakedev0 committed:
fix line length

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into testing-split

cakedev0 commented on sklearn/tree/_utils.pyx:
```diff
@@ -67,6 +69,61 @@ cdef inline float64_t log(float64_t x) noexcept nogil:
+    if start >= end:
+        return 0
+    cdef size_t itemsize = sizeof(array[0])
+    cdef intp_t n_rev = end - start - n
```
Hum... I would prefer not to support invalid inputs like that because none of similar functions do (e.g. `sort`, or `WeightedFenwickTree.reset`, ...).

Maybe I can move this function in `_partitioner.pyx`, it's the only place where it's used, to make it clear it's only meant to be used there and make it more ok to not support invalid inputs?

cakedev0 replied:
> Maybe I can move this function in _partitioner.pyx

Did that

cakedev0 committed:
move swap_array_slices into partitioner.pxy; simplify the logic

cakedev0 committed:
update change log

cakedev0 committed:
KBinsDiscretizer for poisson test

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
fix warning

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
regression test for this PR

cakedev0 committed:
fix regression

ogrisel commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -517,6 +613,84 @@ def test_importances_gini_equal_squared_error():
+    ],
+)
+@pytest.mark.parametrize("min_impurity_decrease_rate", [0.0, 0.1, 0.3])
+def test_impurity_values_on_toy_data(
```
Sorry I forgot about that. Let's remove my test and review #32193 instead.

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into testing-split

cakedev0 committed:
small cleanup

cakedev0 committed:
remove categorical test for now

cakedev0 committed:
continuing removing categorical support

cakedev0 committed:
merged with ogrisel test

cakedev0 committed:
minor formatting

cakedev0 committed:
improvements

cakedev0 committed:
fix & simplify

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into fix/tree_nan_mask

ogrisel committed:
Avoid allocating `(n_samples, n_features)` boolean array.

Co-authored-by: Arthur Lacote <arthur.lcte@gmail.com>

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into fix/tree_nan_mask

cakedev0 committed:
changelog

cakedev0 committed:
changelog: improved wording

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into fix/tree_nan_mask

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into testing-split

cakedev0 committed:
Merge pull request #5 from cakedev0/fix/tree_nan_mask

Fix the NaN-detection bug (#33113)

cakedev0 committed:
renamed changelog

cakedev0 committed:
Apply suggestions from code review

Naming, style, comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

cakedev0 committed:
fix test params

cakedev0 committed:
fix for older numpy version (CI failure)

cakedev0 committed:
testing node values

cakedev0 committed:
small rephrasing

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Revert "TST: compare the single split impurity to pure Python brute force solution"

This reverts commit 5a74902edfc6bbb1689f46c7ff6e840ea6b2d34d.

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into testing-split

cakedev0 committed:
Merge branch 'testing-split' into tree-simpler-missing

cakedev0 committed:
test all cases

cakedev0 committed:
remove changelog from merged PR

cakedev0 committed:
added small todo

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into testing-split

cakedev0 committed:
Merge branch 'testing-split' into tree-simpler-missing

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
[all random seeds] test_split_impurity

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
remove commented import

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

ogrisel reviewed (commented):
I started a review a while ago but did not find the time to complete it. Here are the pending comments. Hopefully I will find the time to resume the review in the coming weeks.

cakedev0 committed:
Improve changelogs

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

adam2392 commented on sklearn/tree/_criterion.pyx:
```diff
@@ -500,11 +424,11 @@ cdef class ClassificationCriterion(Criterion):
         # sum_left from the direction that require the least amount
         # of computations, i.e. from pos to new_pos or from end to new_po.
-        if (new_pos - pos) <= (end_non_missing - new_pos):
+        if (new_pos - pos) <= (self.end - new_pos):
```
For my understanding, what is self.end now? Is it invariant, always tracking the end_non_missing by definition? I.e. the partitioner always keeps missing values on the right, until we do the final evaluation?

adam2392 reviewed (commented):
I need to look at the logic of _partitioned and _splitter, and then I should be done.

adam2392 commented on sklearn/tree/_partitioner.pyx:
```diff
@@ -110,161 +131,126 @@ cdef class DensePartitioner:
-                    max_feature_value = current_feature_value
+            if isnan(current_feature_value):
+                n_missing += 1
+            elif not found_non_missing:
```
Since this is a double negative, I'm having a hard time parsing what the intent is. I think it means "else if we don't find a value that is non missing", i.e. "Elise if we find a value that is not missing"? 

Regardless, probably worth documenting with some comments, or renaming the flag variable name.

adam2392 replied:
Okay I think I get it now. Ah I wonder if `seen_non_missing` makes for a better variable name? 

This is super nit tho, so feel free to take it or leave it.

cakedev0 committed:
Update sklearn/tree/_splitter.pyx

Remove debug const

cakedev0 committed:
rename to seen_non_missing

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
improve doc of SparsePartitioner._partition

cakedev0 commented on sklearn/tree/_partitioner.pyx:
```diff
@@ -448,7 +437,7 @@ cdef class SparsePartitioner:
         else:
             # Data are already split
-            return zero_pos
+            return self.start_positive
```
Improve the doc here and around, this should be much easier to understand now: https://github.com/scikit-learn/scikit-learn/pull/32119/commits/1d9618008ee2cbd57ace645fe5831b991809d121

cakedev0 commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -3043,3 +3045,24 @@ def test_missing_values_and_constant_toy():
+    ids=["multiple-non-missing", "single-non-missing"],
+)
+def test_random_splitter_missing_values_uses_non_missing_min_max(X, y):
+    """Check random-split thresholds are finite when the first sample is missing."""
```
It looks weird because it's a non regression test for the subtle bug I mentioned in another thread 😅

I can explain qickly what was the bug here, would that help?

cakedev0 replied:
Will do :+1:

cakedev0 committed:
remove old guard threshold == INFINITY or threshold == -INFINITY

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 committed:
document regression in test_random_splitter_missing_values_uses_non_missing_min_max

cakedev0 commented on doc/whats_new/upcoming_changes/sklearn.tree/32119.fix.rst:
```diff
@@ -0,0 +1,5 @@
+- Fix calculation of nodes impurity in :class:`tree.DecisionTreeRegressor`
+  and :class:`ensemble.RandomForestRegressor` when missing values are
```
Updated the changelog accordingly.

cakedev0 committed:
pass the SplitRecord to partition_samples_final

adam2392 commented on sklearn/tree/_splitter.pyx:
```diff
@@ -455,19 +448,20 @@ cdef inline int node_split_best(
+                            current_split.threshold == INFINITY or
+                            current_split.threshold == -INFINITY
+                        ):
+                            current_split.threshold = feature_values[p_prev]
```
maybe implicit runtime improvement given this if/else is prolly checked # trees * # nodes * # max_features? haha

adam2392 reviewed (commented):
Mostly LGTM!

lorentzenchr commented on doc/whats_new/upcoming_changes/sklearn.tree/32119.fix.rst:
```diff
@@ -0,0 +1,6 @@
+- Fix calculation of nodes impurity in :class:`tree.DecisionTreeRegressor`,
```
```suggestion
- Fix calculation of node impurity in :class:`tree.DecisionTreeRegressor`,
```
or impurities

lorentzenchr replied:
```suggestion
- Fix calculation of node impurity in :class:`tree.DecisionTreeRegressor`,
```
or node impurities

lorentzenchr commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -115,6 +116,9 @@
 FOREST_CLASSIFIERS_REGRESSORS.update(FOREST_REGRESSORS)
 
+CLF_CRITERIONS = ("gini", "log_loss")
+REG_CRITERIONS = ("squared_error", "absolute_error", "friedman_mse", "poisson")
```
```suggestion
REG_CRITERIONS = ("squared_error", "absolute_error", "poisson")
```
It is deprecated already. Shall we remove it here?

lorentzenchr commented on sklearn/tree/_splitter.pyx:
```diff
@@ -455,19 +447,15 @@ cdef inline int node_split_best(
-                    current_split.n_missing = n_missing
+                    if p == end_non_missing and not missing_go_to_left:
+                        # split with the right node being only the missing values
+                        current_split.threshold = INFINITY
```
Should we have a similar if condition for left has only missing value? Something like
```py
p == 0 and missing_go_to_left
```

cakedev0 replied:
That would test the same split twice (all missing on one side), so we skip this case by doing:
```Python
            if self.missing_go_to_left and p[0] == self.start:
                # skip the missing values up to the first non-missing value:
                p[0] = self.start + self.n_missing
            p[0] += 1
```

lorentzenchr reviewed (commented):
Only _classes.py and some tests left.

cakedev0 committed:
changelog typo

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

cakedev0 committed:
improve doc formulation

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

cakedev0 committed:
Shorter test explanation

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

cakedev0 commented on sklearn/tree/_splitter.pxd:
```diff
@@ -22,8 +22,8 @@ cdef struct SplitRecord:
     float64_t upper_bound     # Upper bound on value of both children for monotonicity
-    uint8_t missing_go_to_left  # Controls if missing values go to the left node.
-    intp_t n_missing            # Number of missing values for the feature being split on
+    bint missing_go_to_left  # Controls if missing values go to the left node.
```
I didn't know, good to know thanks!

cakedev0 committed:
bint -> uint8

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

cakedev0 commented on sklearn/tree/_partitioner.pyx:
```diff
@@ -110,161 +132,127 @@ cdef class DensePartitioner:
+            if self.missing_on_the_left and p[0] == self.start:
+                # skip the missing values up to the first non-missing value:
+                p[0] = self.start + self.n_missing
             p[0] += 1
```
It's set at the end of the function.

cakedev0 committed:
Remove assert

cakedev0 committed:
fix forests test

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

cakedev0 commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -1855,35 +1866,37 @@ def test_missing_values_is_resilient(make_data, Forest):
     n_samples = 300
     expected_score = 0.75
+    # FIXME: this `expected_score` is too high and breaks in some cases
+    # if the seed is set to something else than 0
```
Ok, let's fix it. I used a gap for predictive vs non-predictive instead of an expected score, it's more stable this way and I don't need to use the brier score.

But maybe I can make the test stronger with the brier score, I can give it a try if you want.

cakedev0 committed:
slightly looser gap

cakedev0 committed:
run seed-sensitive tree/forest tests [all random seeds]
  test_forest and test_missing_value_is_predictive
  test_split and test_split_impurity
  test_swap and test_py_swap_array_slices_random
  test_tree and test_regression_tree_missing_values_toy

adam2392 commented:
Let me know if I need to take another look. Otw, I'll let @lorentzenchr and/or @ogrisel converge first.

cakedev0 committed:
added edge cases tests for swap slices

cakedev0 committed:
todo for swap buffer size optim

cakedev0 committed:
Merge remote-tracking branch 'upstream/main' into tree-simpler-missing

ogrisel reviewed (commented):
Overall LGTM besides the following small details but I am still confused (as Christian) by line 450 in `_splitter.pyx`.

ogrisel reviewed (commented):
Assuming the following is correct, let's also expand the docstring of `next_p` to make that more explicit.

Is there a way to test this one way or another? Or is this already tested? If so LGTM.

cakedev0 commented on sklearn/tree/_partitioner.pxd:
```diff
@@ -76,10 +73,13 @@ cdef class DensePartitioner:
     cdef intp_t end
     cdef intp_t n_missing
     cdef const uint8_t[::1] missing_values_in_feature_mask
+    cdef bint missing_on_the_left
```
(wow thx for the very detailed comment)

cakedev0 committed:
Add/improve comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

cakedev0 committed:
remove missing_on_the_left + document a lot next_p

cakedev0 commented on sklearn/tree/_splitter.pyx:
```diff
@@ -455,19 +447,15 @@ cdef inline int node_split_best(
-
-                    current_split.n_missing = n_missing
+                    if p == end_non_missing and not missing_go_to_left:
+                        # split with the right node being only the missing values
```
This is correct, I applied the suggestion.

> Is there a way to test this one way or another? Or is this already tested?

What is "this"?
1. this = "split all missing vs all non-missing"?
2. this = "we don't check all missing on left"

1 is tested in `test_missing_values_and_constant_toy`.

2 seems hard to test...

cakedev0 replied:
I checked and 1 is also tested in `test_split.py`

lorentzenchr reviewed (commented):
This LGTM, just a few nitpicks and the open question from Olivier and me.

cakedev0 committed:
Cleanups & comments

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

ogrisel reviewed (approved):
LGTM.

cakedev0 committed:
y -= min(y)

cakedev0 committed:
Merge branch 'tree-simpler-missing' of github.com:cakedev0/scikit-learn into tree-simpler-missing

cakedev0 committed:
Comment about nan <= inf

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

cakedev0 commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -1829,22 +1833,29 @@ def test_estimators_samples(ForestClass, bootstrap, seed):
+            encode="ordinal", random_state=rng, quantile_method="linear"
+        )
+        y = kbd.fit_transform(y.reshape(-1, 1)).ravel()
+
```
Ok, resolving this thread then :tada:

lorentzenchr reviewed (approved):
@cakedev0 Thank you for this really neat PR. Less code, more and better functionality, that's just 🚀
